### ISSUE TITLE
Hive: SET parameters and configs return PRS error

### ISSUE THREAD

**Issue by antonitto** [sqlfluff/sqlfluff#5713]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Common Hive parameters and configs result in parsing error (see example below). Similar issue was reported before (#1768), but fix https://github.com/sqlfluff/sqlfluff/pull/4763 didn't fix all potential use cases.

I've provided some edge cases below, including `BETWEEN ${two_dates}` and `IN ${list}`.

### Expected Behaviour

Parameters following SET are parsed as standard syntax.

### Observed Behaviour

Example below returns the following error:
```
(sqlfluff-test-2) λ sqlfluff lint test-2.sql --dialect hive
== [test-2.sql] FAIL
L:   1 | P:   5 |  PRS | Line 1, Position 5: Found unparsable section:
                       | 'hive.execution.engine=tez;\nset tez.queue...'
L:  26 | P:   6 |  LXR | Unable to lex characters: '${hivevar:...'
L:  27 | P:  14 |  LXR | Unable to lex characters: '${datafilt...'
WARNING: Parsing errors found and dialect is set to 'hive'. Have you configured your dialect correctly?
All Finished � �!
```

### How to reproduce

```sql
-- hive configs
set hive.execution.engine=tez;
set tez.queue.name=example;
set tez.am.container.reuse.enabled=false;
set tez.runtime.enable.final-merge.in.output=false;
set hive.cbo.enable=true;
set hive.compute.query.using.stats=true;
set hive.vectorized.execution.enabled=false;
set hive.vectorized.execution.reduce.enabled=true;
set hive.fetch.task.conversion=none;
set mapreduce.map.java.opts=-Xmx13107m;
set mapreduce.reduce.memory.mb=16384;
set mapreduce.reduce.java.opts=-Xmx13107m;
set hive.exec.reducers.bytes.per.reducer=256000000;
set mapred.reduce.tasks=600;
set hive.tez.auto.reducer.parallelism=true;
set hive.optimize.skewjoin=true;
set hive.skewjoin.key=500000;
set hive.skewjoin.mapjoin.map.tasks=10000;
set hive.skewjoin.mapjoin.min.split=33554432;

-- hive params
set hivevar:schema = schema_test;
set hivevar:tbl = tbl_test;
set datafilter = 2021 AND 2022;
set list = (1, 2, 3);

SELECT *
FROM ${hivevar:schema}.{tbl}
WHERE year BETWEEN ${datafilter}
    AND col1 IN ${list}
;
```

### Dialect

hive

### Version

sqlfluff==3.0.3
python=3.11

### Configuration

None (default)

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by antonitto** [comment]
Note that a sql file might look like below:
```sql
SELECT *
FROM ${hivevar:schema}.{tbl}
WHERE year BETWEEN ${datafilter}
    AND col1 IN ${list}
;
```

and all configs and params are set in another init file (e.g. `hive -f test.sql -i init.sql`)

**Comment by antonitto** [comment]
Also note that a subquery can be passed as a param:

```sql
SET hivevar:subquery = (
    SELECT DISTINCT client_id in tbl2
);

SELECT *
FROM ${hivevar:schema}.{tbl}
WHERE year BETWEEN ${datafilter}
    AND col1 IN ${list}
    AND client_id IN ${subquery}
;
```
